packages/flutter_tools/lib/src/device.dart

```diff
@@ -334,6 +434,56 @@ abstract class DeviceDiscovery {
   ///
   /// For example, 'windows' or 'linux'.
   List<String> get wellKnownIds;
+
+  bool exactlyMatchesDeviceId(String deviceId, Device device) {
+    final String lowerDeviceId = deviceId.toLowerCase();
+    return device.id.toLowerCase() == lowerDeviceId ||
+        device.name.toLowerCase() == lowerDeviceId;
+  }
+
+  bool startsWithDeviceId(String deviceId, Device device) {
+    final String lowerDeviceId = deviceId.toLowerCase();
+    return device.id.toLowerCase().startsWith(lowerDeviceId) ||
+        device.name.toLowerCase().startsWith(lowerDeviceId);
+  }
+
+  /// Get devices filtered by [filter] that match the given device id/name.
+  ///
+  /// If an exact match is found, return it immediately. Otherwise check all
+  /// devices and return all that begin with the given device id/name.
+  ///
+  /// [waitForDeviceToConnect] is used for iOS device discovery only.
+  Future<DeviceDiscoveryMatchByIdResult> getDevicesById(
+    String deviceId,
+    Logger logger, {
+    DeviceDiscoveryFilter? filter,
+    bool waitForDeviceToConnect = false,
+  }) async {
+    final List<Device> foundDevices = await devices(
+      filter: filter,
+    ).onError((dynamic error, StackTrace stackTrace) {
+      // Fail quietly so other discovers can find matches, even if this one fails.
+      logger.printTrace('Ignored error discovering $deviceId: $error');
+      return <Device>[];
```

Comment:
this code is susceptible to a very obscure gotcha I reported in https://github.com/dart-lang/sdk/issues/51248

```suggestion
    ).then(
      (List<Device> devices) => devices,
      (dynamic error, StackTrace stackTrace) {
        // Fail quietly so other discovers can find matches, even if this one fails.
        logger.printTrace('Ignored error discovering $deviceId: $error');
        return <Device>[];
      }
```

The seemingly silly anonymous closure we pass to `.then()` of `(List<Device> devices) => devices` ensures that the `Future` we call the error handling callback on is of type `List<Device>`, rather than that of a covariant type, such as `List<IosDevice>`.

Reply:
FYI you were not expected to know this, I'm working on landing a lint to catch this.

Reply:
actually, since you're awaiting this, I think you can just use a `try` `catch`:

```dart
    final List<Device> foundDevices;
    try {
      foundDevices = await devices(
        filter: filter,
      );
    } catch (error) {
      // Fail quietly so other discovers can find matches, even if this one fails.
      logger.printTrace('Ignored error discovering $deviceId: $error');
      foundDevices = const <Device>[];
    });
```